### PR TITLE
[Serve][Doc] Add Ray Service CR inspection

### PR DESCRIPTION
**Pull request by sihanwang41**
## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/serve/production-guide/kubernetes.md [hunk 1]
```diff
@@ -28,18 +28,25 @@ Follow the [KubeRay quickstart guide](kuberay-quickstart) to:
 * Prepare a Kubernetes cluster
 * Deploy a KubeRay operator
 
-(serve-deploy-app-on-kuberay)=
-## Deploying a Serve application
+## RayService CR
+Once the KubeRay controller is running, you can manage your Ray Serve application by creating and updating a `RayService` custom resource (CR, here is [example](https://github.com/ray-project/kuberay/blob/release-0.5/ray-operator/config/samples/ray_v1alpha1_rayservice.yaml)).
```
Comment (shrekris-anyscale) [inline doc/source/serve/production-guide/kubernetes.md:32]:
```suggestion
Once the KubeRay controller is running, you can manage your Ray Serve application by creating and updating a `RayService` custom resource (CR).
Here's an [example](https://github.com/ray-project/kuberay/blob/release-0.5/ray-operator/config/samples/ray_v1alpha1_rayservice.yaml)).
```

File: doc/source/serve/production-guide/kubernetes.md [hunk 2]
```diff
@@ -28,18 +28,25 @@ Follow the [KubeRay quickstart guide](kuberay-quickstart) to:
...
+Under the `spec` section in the `RayService` CR, you can find the following fields:
+
+**`serviceUnhealthySecondThreshold`**: represents the threshold in seconds that defines when a service is considered unhealthy. Default is 60 seconds. When the service is unhealthy, the KubeRay Service controller will try to recreate a new cluster and deploy the application to the new cluster.
```
Comment (shrekris-anyscale) [inline doc/source/serve/production-guide/kubernetes.md:35]:
```suggestion
**`serviceUnhealthySecondThreshold`**: the number of seconds before an unhealthy service is marked unhealthy. Default is 60 seconds. When the service is marked unhealthy, the KubeRay Service controller creates a new cluster, deploys the Serve application to it, and then switches traffic to the new cluster.
```

Comment (shrekris-anyscale) [reply doc/source/serve/production-guide/kubernetes.md:35]: Could you give more details on what makes a service unhealthy? E.g. at least one deployment is `UNHEALTHY`? I believe the `app_status` is also relevant here.

File: doc/source/serve/production-guide/kubernetes.md [hunk 3]
```diff
@@ -28,18 +28,25 @@ Follow the [KubeRay quickstart guide](kuberay-quickstart) to:
...
+
+**`deploymentUnhealthySecondThreshold`**: represents the threshold in seconds that defines when the ray dashboard is considered unhealthy. Default is 60 seconds. When the ray dashboard is unhealthy, the KubeRay Service controller will try to recreate a new cluster and deploy the application to the new cluster.
```
Comment (shrekris-anyscale) [inline doc/source/serve/production-guide/kubernetes.md:37]: Based on [this note](https://github.com/shrekris-anyscale/kuberay/blob/af7f80b4b09a01559744ce7f0f262c94242a9be6/docs/guidance/rayservice.md?plain=1#L126), I think `deploymentUnhealthySecondThreshold` limits how long a Serve config should take to be deployed. I don't think it's about the Ray dashboard's health. Could you add more details here?

File: doc/source/serve/production-guide/kubernetes.md [hunk 4]
```diff
@@ -28,18 +28,25 @@ Follow the [KubeRay quickstart guide](kuberay-quickstart) to:
...
 
-Once the KubeRay controller is running, you can manage your Ray Serve application by creating and updating a `RayService` custom resource (CR).
-`RayService` custom resources consist of the following:
-- a `KubeRay` `RayCluster` config defining the cluster that the Serve application runs on.
-- a Ray Serve [config](serve-in-production-config-file) defining the Serve application to run on the cluster.
+**`serveConfig`**: represents the Serve config that will be used to deploy the application. You can use the `--kubernetes-format`/`-k` flag with `serve build` to print the Serve config in a format that can be copy-pasted directly into your [Kubernetes config](serve-in-production-kubernetes). You can paste this config into the `RayService` CR.
+
+**`rayClusterConfig`**: Please refer to [KubeRay configuration](kuberay-config) for more details.
```
Comment (shrekris-anyscale) [inline doc/source/serve/production-guide/kubernetes.md:41]:
```suggestion
**`rayClusterConfig`**: Please refer to [KubeRay configuration](kuberay-config) for more details. This config is identical to the `RayCluster` KubeRay config.
```

File: doc/source/serve/production-guide/kubernetes.md [hunk 5]
```diff
@@ -28,18 +28,25 @@ Follow the [KubeRay quickstart guide](kuberay-quickstart) to:
...
 
 :::{tip}
-You can use the `--kubernetes-format`/`-k` flag with `serve build` to print the Serve config in a format that can be copy-pasted directly into your [Kubernetes config](serve-in-production-kubernetes). You can paste this config into the `RayService` CR.
+To enhance the reliability of your application, particularly when dealing with large dependencies that may require a significant amount of time to download, consider increasing the value of the `serviceUnhealthySecondThreshold` to avoid cluster restart.
```
Comment (shrekris-anyscale) [inline doc/source/serve/production-guide/kubernetes.md:44]: Should they increase the value of `serviceUnhealthySecondThreshold` or `deploymentUnhealthySecondThreshold`?

Comment (shrekris-anyscale) [reply doc/source/serve/production-guide/kubernetes.md:44]: Could you also mention that they should consider installing the dependencies directly into their container?

**Review by angelinalg — APPROVED** [review]

**Review by angelinalg — APPROVED** [review]

**Comment by sihanwang41** [comment]
@edoakes @shrekris-anyscale @zcin please take look,

File: doc/source/serve/production-guide/kubernetes.md [hunk 6]
```diff
@@ -28,18 +28,28 @@ Follow the [KubeRay quickstart guide](kuberay-quickstart) to:
...
 * Deploy a KubeRay operator
 
-(serve-deploy-app-on-kuberay)=
-## Deploying a Serve application
+## Setting up a RayService custom resource (CR)
+Once the KubeRay controller is running, manage your Ray Serve application by creating and updating a `RayService` CR ([example](https://github.com/ray-project/kuberay/blob/release-0.5/ray-operator/config/samples/ray_v1alpha1_rayservice.yaml)).
+
+Under the `spec` section in the `RayService` CR, find the following fields:
+
+**`serviceUnhealthySecondThreshold`**: Represents the threshold in seconds that defines when a service is considered unhealthy (application status is not HEALTHY status). The default is 60 seconds. When the service is unhealthy, the KubeRay Service controller tries to recreate a new cluster and deploy the application to the new cluster.
```
[the remark this answers is not in the document]
Comment (shrekris-anyscale) [reply doc/source/serve/production-guide/kubernetes.md:36]: That makes sense. Is this true also while the app is first deploying?

[the remark this answers is not in the document]
Comment (sihanwang41) [reply doc/source/serve/production-guide/kubernetes.md:36]: No. we respect DeploymentUnhealthySecondThreshold for the first deploy.

**Comment by sihanwang41** [comment]
Hi @akshay-anyscale , can you take a look?

**Comment by akshay-anyscale** [comment]
can you rebase with @angelinalg's latest changes

**Comment by akshay-anyscale** [comment]
while we are changing this,  can you update the example to not use deployment graphs?

**Comment by sihanwang41** [comment]
> while we are changing this, can you update the example to not use deployment graphs?

Currently the deployments are fairly easy. Is there any reasons  to change the example?

**Comment by akshay-anyscale** [comment]
> > while we are changing this, can you update the example to not use deployment graphs?
> 
> Currently the deployments are fairly easy. Is there any reasons to change the example?

It's because deployment graphs are still in alpha and not very completely supported. We're also moving them down in the documentation

File: doc/source/serve/production-guide/kubernetes.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (akshay-anyscale) [reply doc/source/serve/production-guide/kubernetes.md]: just for my clarification - https://github.com/ray-project/ray/issues/35948 . Is this UPDATING status relevant here?

[the remark this answers is not in the document]
Comment (sihanwang41) [reply doc/source/serve/production-guide/kubernetes.md]: it is not relevant. the status here is for application status, not deployment status.

File: doc/source/serve/production-guide/kubernetes.md [hunk 7]
```diff
@@ -28,18 +28,28 @@ Follow the [KubeRay quickstart guide](kuberay-quickstart) to:
...
+
+**`deploymentUnhealthySecondThreshold`**: Represents the number of seconds that the Serve application status can be unavailable before the service is considered unhealthy. The Serve application status is unavailable whenever the Ray dashboard is unavailable. The default is 60 seconds. When the service is unhealthy, the KubeRay Service controller tries to recreate a new cluster and deploy the application to the new cluster.
 
-Once the KubeRay controller is running, you can manage your Ray Serve application by creating and updating a `RayService` custom resource (CR).
-`RayService` custom resources consist of the following:
-- a `KubeRay` `RayCluster` config defining the cluster that the Serve application runs on.
-- a Ray Serve [config](serve-in-production-config-file) defining the Serve application to run on the cluster.
+**`serveConfig`**: Represents the configuration that Ray Serve uses to deploy the application. Use the `--kubernetes-format`/`-k` flag with `serve build` to print the Serve configuration and copy-paste it directly into your [Kubernetes config](serve-in-production-kubernetes) and `RayService` CR.
```
Comment (akshay-anyscale) [inline doc/source/serve/production-guide/kubernetes.md:40]: Maybe add serveConfigv2 here?

Comment (sihanwang41) [reply doc/source/serve/production-guide/kubernetes.md]: good point.

File: doc/source/serve/production-guide/kubernetes.md [hunk 8]
```diff
@@ -28,18 +28,28 @@ Follow the [KubeRay quickstart guide](kuberay-quickstart) to:
...
+
+**`deploymentUnhealthySecondThreshold`**: Represents the number of seconds that the Serve application status can be unavailable before the service is considered unhealthy. The Serve application status is unavailable whenever the Ray dashboard is unavailable. The default is 60 seconds. When the service is unhealthy, the KubeRay Service controller tries to recreate a new cluster and deploy the application to the new cluster.
 
-Once the KubeRay controller is running, you can manage your Ray Serve application by creating and updating a `RayService` custom resource (CR).
-`RayService` custom resources consist of the following:
-- a `KubeRay` `RayCluster` config defining the cluster that the Serve application runs on.
-- a Ray Serve [config](serve-in-production-config-file) defining the Serve application to run on the cluster.
+**`serveConfigV2`**: Represents the configuration that Ray Serve uses to deploy the application. Use the `--kubernetes-format`/`-k` flag with `serve build` to print the Serve configuration and copy-paste it directly into your [Kubernetes config](serve-in-production-kubernetes) and `RayService` CR.
```
Comment (akshay-anyscale) [inline doc/source/serve/production-guide/kubernetes.md:40]: the -k flag is no longer relevant for this config right?

**Review by akshay-anyscale — APPROVED** [review]

**Pull request merged**